lightning/src/util/ser_macros.rs

```diff
@@ -881,19 +921,62 @@ macro_rules! _init_and_read_tlv_stream {
 		$(
 			$crate::_init_tlv_field_var!($field, $fieldty);
 		)*
-
 		$crate::decode_tlv_stream!($reader, {
 			$(($type, $field, $fieldty)),*
 		});
 	}
 }
 
+/// Dummy macro that drops the second argument (which is used by
+/// [`lightning_macros::drop_legacy_field_definition`] to match for legacy fields but isn't needed
+/// in the final code we want to generate).
+#[doc(hidden)]
+#[macro_export]
+macro_rules! _ignore_arg {
+	($field: ident, $fieldty: tt) => {
+		$field
+	};
+}
+
+/// Reads a TLV stream with the given fields to build a struct/enum variant of type `$thing`
+#[doc(hidden)]
+#[macro_export]
+macro_rules! _decode_and_build {
+	($stream: ident, $thing: path, {$(($type: expr, $field: ident, $fieldty: tt)),* $(,)*}) => { {
+		$(
+			$crate::_init_tlv_field_var!($field, $fieldty);
+		)*
+		let tlv_len: $crate::util::ser::BigSize = $crate::util::ser::Readable::read($stream)?;
+		let mut rd = $crate::util::ser::FixedLengthReader::new($stream, tlv_len.0);
+		let rewind = |_, _| { unreachable!() };
+		$crate::_decode_tlv_stream_range!(SKIP_LEGACY, &mut rd, .., rewind, {$(($type, $field, $fieldty)),*});
+		rd.eat_remaining().map_err(|_| $crate::ln::msgs::DecodeError::ShortRead)?;
+		// We mark fields as `mut` as we read them so that legacy post-read actions can modify them
+		// if desired.
+		$(
+			#[allow(unused_mut)]
+			let mut $field = $crate::_init_tlv_based_struct_field!($field, $fieldty);
+		)*
+		$({
+			$crate::_run_legacy_tlv_read_logic!($field, $fieldty);
+		})*
```

Comment:
Not sure I understand why this called both here and from `_run_legacy_tlv_read_logic`.

Reply:
Ah, sorry. I meant from `_decode_tlv_stream_range`, but seems that it is skipped there because we're using `SKIP_LEGACY` here when calling that.